lib/src/client/http2_connection.dart

```diff
@@ -69,25 +73,36 @@ class Http2ClientConnection implements connection.ClientConnection {
 
   ConnectionState get state => _state;
 
+  static const _estimatedRoundTripTime = const Duration(milliseconds: 20);
+
   Future<ClientTransportConnection> connectTransport() async {
-    var socket = await Socket.connect(host, port);
-    if (_state == ConnectionState.shutdown) {
-      socket.destroy();
-      throw _ShutdownException();
-    }
     final securityContext = credentials.securityContext;
-    if (securityContext != null) {
-      socket = await SecureSocket.secure(socket,
-          host: authority,
+    Socket socket;
+    if (securityContext == null) {
+      socket = await Socket.connect(host, port);
+    } else {
+      final socket = await SecureSocket.connect(host, port,
+          supportedProtocols: ['h2'],
           context: securityContext,
           onBadCertificate: _validateBadCertificate);
-      if (_state == ConnectionState.shutdown) {
+      if (socket.selectedProtocol != 'h2') {
         socket.destroy();
-        throw _ShutdownException();
+        throw (TransportException(
+            'Endpoint $host:$port does not support http/2 via ALPN'));
       }
     }
-    socket.done.then((_) => _handleSocketClosed());
-    return ClientTransportConnection.viaSocket(socket);
+
+    final connection = ClientTransportConnection.viaSocket(socket);
+    socket.done.then((_) => _abandonConnection());
+
+    // Give the settings settings-frame a bit of time to arrive.
+    await new Future.delayed(_estimatedRoundTripTime);
```

Comment:
This seems like a smell - can we not be notified somehow when we're ready?

Reply:
Maybe add a comment that explains why we do this, and how it could be better.